src/reducers/steps.js

```diff
@@ -133,72 +206,132 @@ const steps = (state = { language: { constants: "", functions: "", predicates: "
           newRank.set(key, value - 1)
         }
       }
-      return Object.assign({}, state, {
+      return {...state,
         order: [
           ...state.order.filter(id => id !== action.id)
         ],
         allSteps: newSteps,
         rank: newRank,
-      })
+      };
 
-    case INSERT_STEP:
-      return Object.assign({}, state, {
+    case INSERT_STEP: {
+      return {...state,
         order: [
           ...state.order.slice(0, action.position),
           action.id,
           ...state.order.slice(action.position)
         ],
         allSteps: new Map([
           ...state.allSteps,
-          [action.id, {
-            formula: "",
-            rule: "",
-            renaming: "",
-            unifier: "",
-            reference: ""
-          }]
+          [action.id, newStep]
         ]),
         rank: new Map([
           ...state.rank,
           [action.id, action.position]
         ])
-      })
+      }
+    }
 
-    case STEP_UP:
-      newRank = new Map(state.rank);
+    case STEP_UP: {
+      let newRank = new Map(state.rank);
       newRank.set(action.id, action.position - 1);
       newRank.set(state.rank.get(state.order[action.position - 1]), action.position);
-      return Object.assign({}, state, {
+      return {...state,
         order: [
           ...state.order.slice(0, action.position - 1),
           state.order[action.position],
           state.order[action.position - 1],
           ...state.order.slice(action.position + 1)
         ],
         rank: newRank
-      })
+      }
+    }
+      
 
-    case STEP_DOWN:
-      newRank = new Map(state.rank);
+    case STEP_DOWN: {
+      let newRank = new Map(state.rank);
       newRank.set(action.id, action.position + 1);
       newRank.set(state.rank.get(state.order[action.position + 1]), action.position);
-      return Object.assign({}, state, {
+      return {...state,
         order: [
           ...state.order.slice(0, action.position),
           state.order[action.position + 1],
           state.order[action.position],
           ...state.order.slice(action.position + 2)
         ],
         rank: newRank
+      }
+    }
+      
+    case CHANGE_CONST:
+    case CHANGE_FUN:
+    case CHANGE_PRED: {
+      let newAllSteps = new Map([...state.allSteps]);
+      state.allSteps.forEach(function (value, key, map) {
+        let formula = {
+          ...state.allSteps.get(key).formula,
+          error: ""
+        };
+        try {
+          formula.object = parseClause(value.formula.input, getSymbols(language), getFactories(language));
+        } catch (e) {
+          formula.error = e;
+        }
+        newAllSteps = new Map ([...newAllSteps,
+          [key, {
+            ...state.allSteps.get(key),
+            formula
+          }]
+        ]);
       })
+      return {
+        ...state,
+        allSteps: newAllSteps
+      }
+    }
```

Comment:
Priveľakrát vyrábate novú mapu. Okrem toho `state.allSteps.get(key)` je asi to isté, čo `value` a premenné je lepšie nazývať špecifickejšie – napr. v tomto prípade `id` a `step` namiesto `key` a `value`.

Fungovalo by to takto?
```javascript
return {
  ...state,
  allSteps: new Map(state.allSteps.map((step, id) => {
    let object = step.formula.object;
    let error = "";
    try {
      object = parseClause(step.formula.input,
        getSymbols(language), getFactories(language));
    } catch (e) {
      error = e;
    }
    return [
      id,
      {
        ...step,
        formula: { ...step.formula, object, error }
      }
    ]
  })
}
```

Reply:
Aha, tak nie :) `Map` nemá `map()`.
Potom asi takto:
```javascript
      const allSteps = new Map();
      state.allSteps.forEach((step, id) => {
        let object = step.formula.object;
        let error = "";
        try {
          object = parseClause(step.formula.input,
            getSymbols(language), getFactories(language));
        } catch (e) {
          error = e;
        }
        allSteps.set(id,
          {
            ...step,
            formula: { ...step.formula, object, error }
          }
        );
      });
      return { ...state, allSteps };
```